book/src/development/emitting_lints.md

```diff
@@ -0,0 +1,217 @@
+        if some_bar_expr_logic(expr) {
+            span_lint_and_help(
+                cx, // < The context
+                FOO_FUNCTIONS, // < The name of the lint in ALL CAPS
+                expr.span, // < The span to lint
+                "message on why the lint is emitted",
+                None, // < An optional help span (to highlight something in the lint)
+                "message that provides a helpful suggestion",
+            );
+        }
+    }
+}
+```
+
+> Note: According to [the rustc-dev-guide], the message should be matter of fact and avoid
+> capitalization and periods, unless multiple sentences are needed. When code or
+> an identifier must appear in a message or label, it should be surrounded with
+> single grave accents (\`).
+
+## Suggestions: Automatic fixes
+
+Some lints know what to change in order to fix the lint. A real world example, the lint
+[`range_plus`][range_plus_one] lints for ranges where the user wrote `x..y + 1` instead of using an
+[inclusive range][inclusive_range] (`x..=1`). The fix to this lint would be just changing your `x..y + 1` expression
+to `x..=y`, **this is where suggestions come in**.
+
+A suggestion is a change that the lint provides to fix the issue it is linting.
+The output looks something like this (from the example earlier):
+
+```text
+error: an inclusive range would be more readable
+  --> $DIR/range_plus_minus_one.rs:37:14
+   |
+LL |     for _ in 1..1 + 1 {}
+   |              ^^^^^^^^ help: use: `1..=1`
+```
+
+**Not all suggestions are always right**, some of them require human supervision, that's why we have
+[Applicability][applicability].
+
+Applicability indicates confidence in the correctness of the suggestion, some are always right
+(`Applicability::MachineApplicable`), but we use `Applicability::MaybeIncorrect` and other when talking about a lint
+that may be incorrect.
+
+---
+
+The same lint (`foo_functions`) but that emits a suggestion would be something like this:
+
+```rust
+impl<'tcx> LateLintPass<'tcx> for BarExpressions {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>)  {
+        // Imagine that `some_bar_expr_logic` checks for requirements for emitting the lint
+        if some_bar_expr_logic(expr) {
+            span_lint_and_sugg( // < Note this change
+                cx,
+                BAR_EXPRESSIONS,
+                span,
+                "message on why the lint is emitted",
+                "use",
+                "suggestion (don't forget to integrate things from the source, like variable names)", // < Suggestion
+                Applicability::MachineApplicable
+            );
+        }
+    }
+}
+```
+
+Suggestions generally use the [`format!`](format_macro) macro to interpolate the old values with the new ones.
+
+## How to choose between notes, help messages and suggestions
+
+Notes are presented separately from the main lint message, they provide useful information that the user needs to
```

Comment:
I don't think the first part of this sentence is always true?

Reply:
I think it's always true, that's what the documentation says:
> The note message is presented separately from the main lint message and is attached to a specific span